LW-004

WojciechBarwinski opened the pull request:
Zrobiłem to w ten sposób, robiąc z tej DB in-memory singletona. Czy dodanie suffixu do nazwy klasy tutaj do dobra rzecz? 
Nie robiłem żadnego serwisu który by tą DB obsługiwał gdyż:
1. Potrzebuje zatwierdzenia w LW-002 
2. jak rozumiem nie mamy czegoś takiego w taskach, wiec czy modyfikujemy taska czy tworzymy nowego? 

Zastanawiałem się tez nad ID, na razie po za numerkiem 000 dodałem na koncu literę, B-board game, M-minitures, A - accessory, trochę odruchowo, łatwiej zobaczyc czego sie szukac czy coś, ale nie wiem jaki to ma sens. Normalnie DB dawało by "just id" a to co teraz jest literą mogłoby być polem produktu.

WojciechBarwinski committed:
LW-004
Create in-memory DB with starting data.

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockSingleton.java:
```diff
@@ -0,0 +1,41 @@
+
+@Getter
+@Setter
+public class StockSingleton {
```
Zrob prawilnie interfejs StockRepository i jego implementację InMemoryStockRepository

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockSingleton.java:
```diff
@@ -0,0 +1,41 @@
+    private static StockSingleton instance;
+    private List<StockItem> stockItemList;
+
+    private StockSingleton() {
```
Właściwie czemu to nie jest Bean springowy, tylko ręcznie robiony singleton? xd

gestru commented:
W scopie zadania było jeszcze "Przerobić mechanizm wykrywania produktów bez stocków"

WojciechBarwinski committed:
Merge branch 'refs/heads/master' into LW-004

WojciechBarwinski committed:
LW-004.1
Change StockRepositoryImpl, create 2 services for proceed order, change 2 exceptions.

WojciechBarwinski committed:
LW-004.2
Remove BaseRESTController.java
Change in OrderController
Remove "hello world"

WojciechBarwinski committed:
LW-004.3
Add tests StockRepositoryImplTest and StockServiceImplTest

WojciechBarwinski committed:
LW-004.4
Rename some fields and methods

WojciechBarwinski committed:
LW-004.5
Remove StockService, refactor OrderServiceImpl, add test.
Overall code refactoring

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/controllers/ApplicationExceptionHandlerController.java:
```diff
@@ -17,8 +18,15 @@ public class ApplicationExceptionHandlerController {
+        return new ErrorResponse<>(exception.getMessage(), exception.getInsufficientStock());
+    }
+
+    @ResponseStatus(HttpStatus.UNPROCESSABLE_ENTITY) // 422
```
To jest ciekawy case, omówimy sobie w weekend ;)

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/dtos/InsufficientStockDTO.java:
```diff
@@ -0,0 +1,4 @@
+package com.wojciechbarwinski.demo.legendary_warehouse.dtos;
+
+public record InsufficientStockDTO(String productId, int expectedQuantity, int stockQuantity) {
```
Ze skrótami czy akronimami na dwoje babka wróżyła - jedni wolą XxxYyyDTO, inni XxxYyyDto jakby co ;) to samo było z RESTController (aczkolwiek tutaj raczej spotyka się RestController)

WojciechBarwinski replied:
oki, rozumiem
mi jakoś łatwiej z DTO, łatwiej rzuca sie w oczy i od razu wiesz co to. Ale nie problem zmienić

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/exceptions/InsufficientStockException.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+public class InsufficientStockException extends ApplicationException {
+    private final List<InsufficientStockDTO> insufficientStock;
```
Jak mamy listę (mnogą liczbę elementów) dobrze jest to odzwierciedlić w nazwie pola/zmiennej

WojciechBarwinski replied:
racja, ale tu z chęcią bym pogadał bo jakoś ogólnie mam problemy z nazywaniem list xd

gestru replied:
Trzeba zazcznaczyć, że liczba mnoga, ale na którym rzeczowniku to już kwestia konwencji ;)

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/exceptions/InsufficientStockException.java:
```diff
@@ -0,0 +1,18 @@
+    private final String message;
+
+    public InsufficientStockException(String message, List<InsufficientStockDTO> insufficientStock) {
+        this.message = message;
```
Tutaj pytanie, czy na pewno chcemy dawać możliwość manipulacją message temu, kto programuje rzucenie tego wyjątku. Dość często w parametrze konstruktora nie dajesz message, tylko ustawiasz go na sztywno w body konstruktora. W ten sposób zapewniasz unifikację wiadomości dla danego typu wyjątku (tj. każdy rzucony wyjątek tego typu będzie miał podobny message)

WojciechBarwinski replied:
nigdy o tym nie myslałem ale w sumie to faktycznie ma sens. 
Nie trzeba sie zastanawiać jak treścią 
a jak 1 wyjątek raz zwracał by wiadomośc zxc a raz qwe to byłaby lipa

gestru replied:
No to zmieniamy :D

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/exceptions/ProductNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+    private final String message;
+
+    public ProductNotFoundException(String message, List<String> missingProductList) {
+        this.message = message;
```
Tutaj pytanie, czy na pewno chcemy dawać możliwość manipulacją message temu, kto programuje rzucenie tego wyjątku. Dość często w parametrze konstruktora nie dajesz message, tylko ustawiasz go na sztywno w body konstruktora. W ten sposób zapewniasz unifikację wiadomości dla danego typu wyjątku (tj. każdy rzucony wyjątek tego typu będzie miał podobny message)

gestru replied:
No to zmieniamy :D

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        removeQuantityOfItemsFromStock(order);
+    }
+
+    private void checkIfProductsInStockAreAvailable(OrderDTO order) {
```
To sobie omówimy jak się zdzwonimy :D

WojciechBarwinski replied:
oki, jestem ciekaw o co chodzi. Bo pamiętam że sam mi o takich metodach mowiłeś

gestru replied:
Przerobic na jeden strzał do  repo

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/Product.java:
```diff
@@ -0,0 +1,16 @@
+@Getter
+@Setter
+@AllArgsConstructor
+@EqualsAndHashCode
```
Swoją drogą... nie pamiętam już, ale czy czasem Lombokowa @Data nie ogra wszystkich tych adnotacji? Ewentualnie zrobienie z tego recordu

WojciechBarwinski replied:
nie daje tego allArg a jakis RequiredArgsConstructor
swoją drogę tu miałym pytanie
czy powinienm od razu dla wygody wrzucac @Data, allArg, noArg czy jednak tylko jak bedzie potrzeba

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockRepositoryImpl.java:
```diff
@@ -0,0 +1,47 @@
+    private Map<String, StockItem> productsMap;
+
+    private StockRepositoryImpl() {
+        productsMap = new ConcurrentHashMap<>();
```
Czemu Concurrent? :D

WojciechBarwinski replied:
specjalnie to tak jakos zaznaczyłem bo... a co jak beda 2 zamowienia na raz sie procedowały i miedzy sprawdzeniem a usunięciem z magazynu, inne zamowienie dokona sprawdzenia? 
Chodziłomi aby w ogóle ten watek ... wielowątkowy poruszyć i czy mógłbyś mi cokolwiek z nim wyjaśnić.

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+    @InjectMocks
+    private OrderServiceImpl orderService;
+
+    @Test
```
Testy sobie też przejdziemy na żywca ;)

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@SpringBootTest()
```
To podnosi kontekst springowy, a więc nie jest to test jednostkowy

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+    private StockRepository stockRepository;
+
+    @InjectMocks
+    private OrderServiceImpl orderService;
```
Alternatywnie:
 private OrderServiceImpl orderService = new OrderServiceImpl(stockRepository);

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+        );
+        OrderDTO order = new OrderDTO(null, orderLineDTOS);
+
+        when(stockRepository.findById(wrongId1)).thenReturn(Optional.empty());
```
Mockito dajee nam coś takiego jak "any()" - wildcard na parametr

gestru replied:
Ewentualnie możemy to olać i zdefiniować tylko dla jednego id - to też nam zaspokoi przypadek testowy

Na pewno powinniśmy mieć tylko jedno wywołanie metody, którą testujemy

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+
+        //when
+        ProductNotFoundException exception1 = assertThrows(ProductNotFoundException.class, () -> orderService.proceedOrder(order));
+        ProductNotFoundException exception2 = assertThrows(ProductNotFoundException.class, () -> orderService.proceedOrder(order));
```
Ta linijka nie ma sensu - w ramach metody lecimy po obu produktach i dopisujemy ich idki do listy w wyjątku

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+        //then
+        assertEquals("Some of the ordered products are not available in our database.", exception1.getMessage());
+        assertTrue(exception1.getMissingProducts().contains(wrongId1));
+        assertTrue(exception2.getMissingProducts().contains(wrongId2));
```
exception1

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/services/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Test
+    void shouldThrowInsufficientStockExceptionWhenOrderQuantityExceedsStock() {
```
Tutaj można by zrobić, żeby jeden miał odpowiedni stock i sprawdzić, czy ten przypadek na pewno NIE występuje w wyjątku

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockRepositoryImplTest.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Optional;
+
+
+@SpringBootTest()
```
To nie jest ten jednostkowy

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockRepositoryImpl.java:
```diff
@@ -0,0 +1,47 @@
+@Getter
+@Setter
+@Repository
+public class StockRepositoryImpl implements StockRepository {
```
Rename na InMemoryStockRepository

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockRepositoryImplTest.java:
```diff
@@ -0,0 +1,61 @@
+@SpringBootTest()
+class StockRepositoryImplTest {
+
+    @Autowired
```
Tego nie chcemy - chcemy tu mieć nowy obiekt

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockRepositoryImplTest.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Test
+    public void shouldFindByIdWithCorrectAndIncorrectId() {
```
To powinny być osobne testy

WojciechBarwinski committed:
LW-004.6
All require changes are done

gestru commented on src/test/java/com/wojciechbarwinski/demo/legendary_warehouse/stock/StockRepositoryImplTest.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Set;
+
+
+class StockRepositoryImplTest {
```
Jak robimy rename na klasie warto pamiętać o sprawdzeniu, czy intellijka nam zmieniła również nazwę klasy testu ;)

WojciechBarwinski committed:
LW-004.7
All require changes are done

WojciechBarwinski committed:
LW-004.8
All require changes are done

WojciechBarwinski committed:
LW-004.9
change message in exceptions
